server/jamovi/server/dataset/duckdataset.py

```diff
@@ -172,26 +181,58 @@ def __iter__(self) -> Iterator[DuckColumn]:
         for column in self._columns_by_index:
             yield column
 
-    def set_value(self, row: int, column: int, value, initing=False):
+    def set_value(self, row: int, column_index: int, value, initing=False):
         """set a value in the data set"""
-        col = self._columns_by_index[column]
-
-        if isinstance(value, str) and col.data_type is DataType.TEXT and col.has_levels:
+        # Question: Did you want the signature to be changed to take row and column as a tuple?
+        col = self._columns_by_index[column_index]
+        value = self._convert_value_to_raw(value, col)
+        self._write_buffer.stage(CellCoordinate(row_index=row, col_iid=col.iid), value=value)
+
+    def set_values(self, values:dict[CellCoordinate, CellValue]):
+        self._set_values(values, index_name="index")
+
+    def _set_values(self, values:dict[CellCoordinate, CellValue], index_name):
+        
+        # Values written one column at a time base on the assumption that cells 
+        # to be set are not neccessarily adjacent.
+         
+        # Note: Could be improved by updating groups of columns that have 
+        # updates for all the same rows as each other
+        col_iids = set([t[1] for t in values.keys()])
+        for col_i in col_iids:
+            col_updates:dict[int,CellValue] = {index:value for (index, col_j), value in values.items() if col_i==col_j}
+            self._update_column_values(col_i, col_updates, index_name)
```

Comment:
i think i'd push this logic into the write buffer class ... `DuckDataset` is already pretty complicated ... so i'd have the write buffer do this. i.e. on a commit, the write buffer calls `set_values()` multiple times, once per column.

Reply:
...and you still want `set_values_method` to be injected into the buffer on creation? 

```py
class DataWriteBuffer:
    ...

    def __init__(
            self, 
            set_values_method: Callable[[dict[CellCoordinate, CellValue]], None],
            max_items: int = 50
            ):
        ...
```

I can do that

Reply:
> ...and you still want set_values_method to be injected into the buffer on creation?

*shrugs* ... not sure what the alternative would be